src/hotspot/share/jeandle/jeandleAbstractInterpreter.hpp

```diff
@@ -81,18 +84,20 @@ class JeandleVMState : public JeandleCompilationResourceObj {
   llvm::Value* dpop() { return pop(BasicType::T_DOUBLE); }
 
   // Untyped manipulation (for dup_x1, etc.)
-  void raw_push(llvm::Value* t) { _stack.push_back(t); }
-  llvm::Value* raw_pop() { llvm::Value* v = _stack.back(); _stack.pop_back(); return v; }
+  void raw_push(TypedValue tv) { _stack.push_back(tv); }
+  TypedValue raw_pop() { TypedValue v = _stack.back(); _stack.pop_back(); return v; }
 
   // Local variables operations:
 
   size_t locals_size() const { return _locals.size(); }
   size_t max_locals() const { return _locals.size(); }
 
-  void invalidate_local(int index) { _locals[index] = nullptr; }
+  void invalidate_local(int index) { _locals[index] = TypedValue(T_VOID, nullptr); }
 
-  llvm::Value* locals_at(int index) { return _locals[index]; }
-  void set_locals_at(int index, llvm::Value* value) { _locals[index] = value; }
+  llvm::Value* locals_at(int index) { return _locals[index].value(); }
+  BasicType locals_type_at(int index) { return _locals[index].actual_type(); }
+  BasicType locals_computational_type_at(int index) { return _locals[index].computational_type(); }
+  void set_locals_at(int index, BasicType bt, llvm::Value* value) { _locals[index] = TypedValue(bt, value); }
```

Comment:
```suggestion
  void set_locals_at(int index, TypedValue value) { _locals[index] = value; }
```
This is a more clear version.